frontend/apps/reader/modules/readerlink.lua

```diff
@@ -233,15 +233,25 @@ end
 function ReaderLink:onGesture() end
 
 function ReaderLink:registerKeyEvents()
-    if Device:hasKeys() and not Device:hasFiveWay() then
+    if Device:hasScreenKB() or Device:hasSymKey() then
+        self.key_events.GotoSelectedPageLink = { { "Press" }, event = "GotoSelectedPageLink" }
+        if Device:hasKeyboard() then
+            self.key_events.AddCurrentLocationToStack = { { "Shift", "Down" } }
+            self.key_events.SelectNextPageLink = { { "Shift", "LPgFwd" }, event = "SelectNextPageLink" }
+            self.key_events.SelectPrevPageLink = { { "Shift", "LPgBack" }, event = "SelectPrevPageLink" }
+        else
+            self.key_events.AddCurrentLocationToStack = { { "ScreenKB", "Down" } }
```

Comment:
could you change this to `ScreenKB` and `Press` (instead of `Down`) there is something I want to try and if successful would go on KB+Down. Also, any idea why this doesn't show a notification when one uses it? normally at the top of the screen a little box appears saying, "added to stack or whatever" (yes all notifications under Screen menu are ON).

Reply:
It looks like notification true or false is passed as an argument. DIspatcher sets arg=true.

Reply:
I’ve already tried that and still won’t show it… not sure why

Reply:
Hm yes, the true comes through if you add `args=true` but it doesn't seem to show the notification. You say it works from Dispatcher?

Reply:
Yes.

Reply:
I think @NiLuJe last touched notifications. I don't know much about them otoh.

Reply:
Because it's `args`, plural, for `InputContainer`'s `key_events` ;).

Reply:
It's the same both with arg or args, i.e does not work

Reply:
Right, take two then ;D.

IIRC, Dispatcher brackets its calls in a way that ensure the notification source is set as Dispatcher.

Here, that's not the case, so the defaults (or the last set source if there's some misbehaving code) is used, and it's likely to be masked out in the default settings, so the notification doesn't show.

Reply:
> the defaults

Being `SOURCE_OTHER`, which is described as `all other sources (e.g. keyboard)`, which is indeed masked out unless you explicitly request notifications from *all* sources.

TL;DR: behaving as expected ;).

Reply:
Right, so how do we set a source and/or make it show the notification?

Reply:
We... don't because it's behaving as expected and keyboard shortcuts shouldn't display notification unless the user asks for it?

Reply:
Yes, but this is 'different'...? I need the notification to be shown so it provides feedback on NT kindles, otherwise it appears as if nothing was happening when that would just not the case.

Essentially it should work same as if it was coming from dispatcher.

Reply:
Then you need a custom event that either wraps the current one in `Notify:setNotifySource(NOTIFICATION.SOURCE_ALWAYS_SHOW)` <-> `Notification:resetNotifySource()` or reimplement the current one except that it passes `NOTIFICATION.SOURCE_ALWAYS_SHOW` as the second arg to `Notify`.

Reply:
Slightly unrelated question, do you have this SOED dictionary @NiLuJe? (mentioned here https://www.mobileread.com/forums/showthread.php?t=347490)

It would be a shame if someone were to accidentally upload it here for 12hrs tops, wink. And then deleted it for ever once they realised their mistake.